src/sage/rings/species.py

```diff
@@ -0,0 +1,1250 @@
+            sage: C = At(G, {1: [1,2,5,6], 2: [3,4,7,8,9,10]})
+            sage: A != B
+            True
+            sage: A == C
+            True
+        """
+        if parent(self) != parent(other):
+            return False
+        # Check if multicardinalities match
+        if self._mc != other._mc:
+            return False
+        # If they do, construct the mapping between the groups
+        selflist, otherlist = [], []
+        for i in range(self.parent()._k):
+            if len(self._dompart[i]) != len(other._dompart[i]):
+                return False
+            selflist.extend(self._dompart[i])
+            otherlist.extend(other._dompart[i])
+        mapping = libgap.MappingPermListList(selflist, otherlist)
+        G = PermutationGroup(gap_group=libgap.ConjugateGroup(self._dis._C, mapping),
+                             domain=self._dis._C.domain())
+        # The conjugated group must be exactly equal to the other group
+        return G == other._dis._C
+
+    def _repr_(self):
+        r"""
+        Return a string representation of ``self``.
+
+        TESTS::
+
+            sage: At = AtomicSpecies(2)
+            sage: G = PermutationGroup([[(1,2),(3,4),(5,6),(7,8,9,10)]]); G
+            Permutation Group with generators [(1,2)(3,4)(5,6)(7,8,9,10)]
+            sage: A = At(G, {1: [1,2,3,4], 2: [5,6,7,8,9,10]}); A
+            {((1,2,3,4)(5,6)(7,8)(9,10),): ((5, 6, 7, 8), (9, 10, 1, 2, 3, 4))}
+        """
+        return "{" + f"{self._dis}: {self._dompart}" + "}"
+
+
+class AtomicSpecies(UniqueRepresentation, Parent, ElementCache):
+    @staticmethod
+    def __classcall__(cls, k, singleton_names=None):
+        """
+        Normalize the arguments.
+
+        TESTS::
+
+            sage: A1 = AtomicSpecies(1, "X")
+            sage: A2 = AtomicSpecies(1)
+            sage: A3 = AtomicSpecies(["X", "Y"])
+            sage: A4 = AtomicSpecies(2, ["X", "Y"])
+            sage: A1 is A2
+            False
+            sage: A3 is A4
+            True
+        """
+        if singleton_names is None:
+            if k in ZZ:
+                k = ZZ(k)
+            else:
+                singleton_names = tuple(k)
+                k = len(singleton_names)
+        else:
+            k = ZZ(k)
+            singleton_names = tuple(singleton_names)
+
+        if (singleton_names is not None
+            and (len(singleton_names) != k
+                 or not all(isinstance(X, str) for X in singleton_names))):
+            raise ValueError(f"singleton_names must be a tuple of {k} strings")
+
+        return super().__classcall__(cls, k, singleton_names)
+
+    def __init__(self, k, singleton_names):
+        r"""
+        Infinite set of `k`-variate atomic species graded by
+        integer vectors of length `k`.
+
+        INPUT:
+
+        - ``k`` -- a non-negative integer, or an iterable of ``k`` strings
+        - ``singleton_names`` -- an iterable of ``k`` strings or ``None``
```

Comment:
I just realised that we can make our life much easier if we require `singleton_names` and do not allow `AtomicSpecies(k)`.  This would also be in line with the polynomial rings in sage.  Thus::
```suggestion
        - ``names`` -- an iterable of ``k`` strings
```